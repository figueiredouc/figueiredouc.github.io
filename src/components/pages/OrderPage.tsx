import React from 'react';
import { Typography, Container, Divider, TableHead } from '@material-ui/core';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import {
  Theme,
  createStyles,
  makeStyles,
  withStyles,
  fade,
} from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableRow from '@material-ui/core/TableRow';
import WarningIcon from '@material-ui/icons/Warning';
import isEmpty from 'lodash/isEmpty';
import pick from 'lodash/pick';
import AddRemoveShopBtns from 'components/shared/AddRemoveShopBtns';
import AppPage from 'components/shared/AppPage';
import { useAuthContext } from 'context/AuthContext';
import { useShopCartContext } from 'context/ShopCartContext';
import { useOrder } from 'hooks';

const TAX_RATE = 0.07;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      height: 'inherit',
      flexGrow: 1,
      paddingTop: theme.spacing(8),
      paddingLeft: 0,
      paddingRight: 0,
      '& .MuiButton-root': {
        backgroundColor: '#CCCCCC',
        margin: theme.spacing(2),
      },
      '& .MuiButton-root:hover': {
        backgroundColor: fade('#660000', 0.3),
      },
    },
    title: { color: '#033860' },
    divider: { backgroundColor: '#B80C09' },
    item: { display: 'flex' },
    table: { minWidth: 700 },
    addRemoveBtns: { display: 'flex', justifyContent: 'space-evenly' },
    avatarProduct: { display: 'flex', justifyContent: 'center' },
    productDetails: { display: 'flex', alignItems: 'center' },
    productQuantity: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'flex-end',
    },
    itemName: { paddingLeft: theme.spacing(2) },
    requestBtns: {
      display: 'flex',
      justifyContent: 'flex-end',
      alignItems: 'center',
    },
    warningMessage: { padding: theme.spacing(2) },
    emptyCart: { display: 'flex', alignItems: 'center' },
  })
);

const StyledTableRow = withStyles((theme: Theme) =>
  createStyles({
    root: {
      '&:nth-of-type(odd)': {
        backgroundColor: theme.palette.action.hover,
      },
    },
  })
)(TableRow);

const ccyFormat = (num: number) => `${num.toFixed(2)}`;

const OrderPage = () => {
  const classes = useStyles();
  const { loggedUser } = useAuthContext();
  const { shopCart, handleCart, shopCartTotalPrices } = useShopCartContext();
  const taxes = TAX_RATE * shopCartTotalPrices.totalPrice;
  const { createOrder } = useOrder();

  const handleRequestOrder = (kind: number) => {
    if (loggedUser) {
      const { token } = loggedUser;

      const products = shopCart.map((item) => pick(item, ['id', 'quantity']));

      if (token) createOrder({ kind, products, token });
    }
  };

  return (
    <AppPage>
      <Container className={classes.root}>
        <Typography className={classes.title} variant="h5">
          Encomenda
        </Typography>
        <Divider className={classes.divider} />
        {isEmpty(shopCart) ? (
          <div className={classes.emptyCart}>
            <WarningIcon />
            <Typography className={classes.warningMessage} variant="body2">
              Não existem produtos no carrinho!
            </Typography>
          </div>
        ) : (
          <TableContainer>
            <Table aria-label="spanning table" className={classes.table}>
              <TableHead>
                <TableRow>
                  <TableCell align="center" colSpan={3}>
                    Detalhes
                  </TableCell>
                  <TableCell align="right">Preço</TableCell>
                </TableRow>
                <TableRow>
                  <TableCell>Produto</TableCell>
                  <TableCell align="right">Qtd.</TableCell>
                  <TableCell align="right">Unidade</TableCell>
                  <TableCell align="right">Soma</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {shopCart.map((item) => (
                  <StyledTableRow key={item.id}>
                    <TableCell>
                      <div className={classes.productDetails}>
                        <Avatar alt={item.name} src={item.photo} />
                        <Typography
                          className={classes.itemName}
                          variant="body2"
                        >
                          {item.name}
                        </Typography>
                      </div>
                      <Typography noWrap color="textSecondary" variant="body2">
                        {item.description}
                      </Typography>
                    </TableCell>
                    <TableCell align="right">
                      <div className={classes.productQuantity}>
                        <AddRemoveShopBtns
                          handleCart={handleCart}
                          item={item}
                        />
                      </div>
                    </TableCell>
                    <TableCell align="right">{item.price}</TableCell>
                    <TableCell align="right">
                      {ccyFormat(item.price * item.quantity)}
                    </TableCell>
                  </StyledTableRow>
                ))}

                <TableRow>
                  <TableCell rowSpan={3} />
                  <TableCell colSpan={2}>Subtotal</TableCell>
                  <TableCell align="right">
                    {ccyFormat(shopCartTotalPrices.totalPrice)}
                  </TableCell>
                </TableRow>
                <TableRow>
                  <TableCell>Tax</TableCell>
                  <TableCell align="right">{`${(TAX_RATE * 100).toFixed(
                    0
                  )} %`}</TableCell>
                  <TableCell align="right">{ccyFormat(taxes)}</TableCell>
                </TableRow>
                <TableRow>
                  <TableCell colSpan={2}>Total</TableCell>
                  <TableCell align="right">
                    {ccyFormat(shopCartTotalPrices.totalPriceBeforeTax)}
                  </TableCell>
                </TableRow>
              </TableBody>
            </Table>
          </TableContainer>
        )}
        {!isEmpty(shopCart) && (
          <div className={classes.requestBtns}>
            {loggedUser ? (
              <>
                <Button onClick={() => handleRequestOrder(1)}>
                  Pedir orçamento
                </Button>
                <Button onClick={() => handleRequestOrder(2)}>
                  Encomendar
                </Button>
              </>
            ) : (
              <>
                <WarningIcon />
                <Typography className={classes.warningMessage} variant="body2">
                  Por favor inicie sessão para proceder com a encomenda!
                </Typography>
              </>
            )}
          </div>
        )}
      </Container>
    </AppPage>
  );
};

export default OrderPage;
