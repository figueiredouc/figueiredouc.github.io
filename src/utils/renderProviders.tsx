/* eslint-disable */
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router-dom'; // 4.0.0
import { render } from '@testing-library/react';
import { AuthContext, AuthStore } from 'context/AuthContext';
import {
  BreadcrumbsContext,
  BreadcrumbsStore,
} from 'context/BreadcrumbsContext';
import { ShopCartContext, ShopCartStore } from 'context/ShopCartContext';
import {
  linkedBreadcrumbs as defaultLinkedB,
  activeBreadcrumb as defaultActiveB,
} from '__mocks__/breadcrumbs';
import { shopCart as defaultShopCart } from '__mocks__/products';

export const defaultStore = {
  AuthStore,
  ShopCartStore,
  BreadcrumbsStore,
};

const customRender = (ui: any, { store = defaultStore, ...options } = {}) => {
  const CombinedProviders = ({ children }: any) => {
    return (
      <AuthContext.Provider value={store.AuthStore()}>
        <ShopCartContext.Provider value={store.ShopCartStore()}>
          <BreadcrumbsContext.Provider value={store.BreadcrumbsStore()}>
            <MemoryRouter>{children}</MemoryRouter>
          </BreadcrumbsContext.Provider>
        </ShopCartContext.Provider>
      </AuthContext.Provider>
    );
  };

  return render(ui, { wrapper: CombinedProviders, ...options });
};

export const customStore = ({
  isLoadingAuthStore = false,
  getLoggedUser = jest.fn(),
  createUser = jest.fn(),
  loggedUser = { user: 'xpto', id: 1, token: '123' },
  shopCart = defaultShopCart,
  shopCartTotalPrices = {
    totalPrice: '20',
    totalPriceBeforeTax: '40',
  },
  linkedBreadcrumbs = defaultLinkedB,
  activeBreadcrumb = defaultActiveB,
  useDefaultAuthStore = false,
  useDefaultShopCartStore = false,
  useDefaultBreadcrumbsStore = false,
} = {}) => {
  const newAuthStore = () => ({
    isLoadingUser: isLoadingAuthStore,
    getLoggedUser,
    loggedUser,
    createUser,
  });

  const newShopCartStore = () => ({
    shopCart,
    shopCartTotalPrices,
  });

  const newBreadcrumbsStore = () => ({
    linkedBreadcrumbs,
    activeBreadcrumb,
  });

  return {
    ...defaultStore,
    AuthStore: useDefaultAuthStore ? defaultStore.AuthStore : newAuthStore,
    ShopCartStore: useDefaultShopCartStore
      ? defaultStore.ShopCartStore
      : newShopCartStore,
    BreadcrumbsStore: useDefaultBreadcrumbsStore
      ? defaultStore.BreadcrumbsStore
      : newBreadcrumbsStore,
  };
};

export * from '@testing-library/react';
export { customRender as render };
