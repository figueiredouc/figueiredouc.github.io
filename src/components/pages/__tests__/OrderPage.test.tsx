import React from 'react';
import { render } from 'utils/renderProviders';
import OrderPage from 'components/pages/OrderPage';

describe('renders OrderPage', () => {
  it('should match snapshot', () => {
    const { container } = render(<OrderPage />);

    expect(container).toMatchSnapshot();
  });
});
