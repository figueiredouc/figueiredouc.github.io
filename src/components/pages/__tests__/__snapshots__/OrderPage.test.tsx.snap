// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders OrderPage should match snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-17"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-appBar-19 MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular makeStyles-header-20 MuiToolbar-gutters"
      >
        <div
          class="makeStyles-logoContainer-21"
        >
          <button
            aria-controls="long-menu"
            aria-haspopup="true"
            aria-label="more"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <img
            alt="cmai"
            height="50px"
            src="logo.png"
          />
        </div>
        <div
          class="makeStyles-flex-27"
        >
          <div
            class="makeStyles-search-29"
          >
            <div
              class="makeStyles-searchIcon-30"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root makeStyles-inputRoot-31"
            >
              <input
                aria-label="search"
                class="MuiInputBase-input makeStyles-inputInput-32"
                placeholder="Search…"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                  />
                </svg>
              </span>
              Iniciar Sessão
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
  <div
    class="makeStyles-dividerMargin-23"
  >
    <hr
      class="MuiDivider-root makeStyles-divider-22"
    />
  </div>
  <div
    class="MuiContainer-root makeStyles-container-14 MuiContainer-maxWidthLg"
  >
    <div
      class="MuiContainer-root makeStyles-root-1 MuiContainer-maxWidthLg"
    >
      <h5
        class="MuiTypography-root makeStyles-title-2 MuiTypography-h5"
      >
        Encomenda
      </h5>
      <hr
        class="MuiDivider-root makeStyles-divider-3"
      />
      <div
        class="makeStyles-emptyCart-13"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
        <p
          class="MuiTypography-root makeStyles-warningMessage-12 MuiTypography-body2"
        >
          Não existem produtos no carrinho!
        </p>
      </div>
    </div>
  </div>
</div>
`;
